tests/unit/test_webhook.py

```diff
@@ -401,26 +411,54 @@ def test_signature_check(self):
+        mock_mr.return_value.status_code == 200
+        json_event, signature = event_data('private_issue_opened.json')
+        payload = json.loads(json_event)
+        issue = WebHookIssue.from_dict(payload)
+        issue.close_private_issue()
+        assert issue.state == 'closed'
+
+    @patch('webcompat.webhooks.model.make_request')
+    def test_WebHookIssue_close_private_issue_fails(self, mock_mr):
+        """Test issue state after a simulated GitHub failure."""
+        mock_mr.side_effect = HTTPError()
+        mock_mr.status_code = 400
+        json_event, signature = event_data('private_issue_opened.json')
+        payload = json.loads(json_event)
+        issue = WebHookIssue.from_dict(payload)
+        with pytest.raises(HTTPError):
+            issue.close_private_issue()
+        assert issue.state == 'open'
+
+    @patch('webcompat.webhooks.model.make_request')
+    def test_WebHookIssue_tag_as_public(self, mock_mr):
+        """Test tagging an issue as public."""
         # A 200 response
         json_event, signature = event_data('new_event_valid.json')
         payload = json.loads(json_event)
-        issue_info = helpers.get_issue_info(payload)
-        with patch('webcompat.webhooks.helpers.proxy_request') as proxy:
-            proxy.return_value.status_code = 200
-            response = helpers.tag_new_public_issue(issue_info)
-            self.assertEqual(response.status_code, 200)
-            # A 401 response
-            proxy.return_value.status_code = 401
-            proxy.return_value.content = '{"message":"Bad credentials","documentation_url":"https://developer.github.com/v3"}'  # noqa
-            with patch.dict('webcompat.webhooks.helpers.app.config',
-                            {'OAUTH_TOKEN': ''}):
-                response = helpers.tag_new_public_issue(issue_info)
-                self.assertEqual(response.status_code, 401)
-                self.assertTrue('Bad credentials' in response.content)
-
-    @patch('webcompat.webhooks.helpers.tag_new_public_issue')
+        mock_mr.return_value = MagicMock(status_code=200, spec=Response)
+        issue = WebHookIssue.from_dict(payload)
+        try:
+            issue.tag_as_public()
+            # did the milestone get set, and were labels updated?
+            method, uri, data = mock_mr.call_args[0]
+            assert issue.milestone == webcompat.app.config['STATUSES']['needstriage']['id']  # noqa
+            assert all(x in data['labels'] for x in
+                       ['browser-firefox', 'engine-gecko'])
+        except HTTPError:
+            pytest.fail("Unexpected requests.exceptions.HTTPError!")
+        # A 401 response
+        with patch.dict('webcompat.webhooks.helpers.app.config',
+                        {'OAUTH_TOKEN': ''}):
+            mock_mr.side_effect = HTTPError()
+            mock_mr.status_code = 401
+            with pytest.raises(HTTPError):
+                issue.tag_as_public()
```

Comment:
@miketaylr 

I was puzzled by this test because it seemed very complicated. 

When I do not understand I try to go back to the code itself. So let's check what we need to test `tag_as_public`

https://github.com/webcompat/webcompat.com/blob/98412a9a5239724feed3c73a0d707c527a04493d/webcompat/webhooks/model.py#L152-L178

1. it doesn't take any parameters so that's one thing less to test on the input side.
2. it has variability inside the function such as labels and milestones. (All of these are tested elsewhere)
3. it creates a `payload_request` which should probably be named `request_payload` (nit) and a `path` 
4. It returns an HTTP response

So there are only 6 tests for this:

1. assert that the request payload is a `dict`
2. assert that the request payload keys are `label` and `milestone`
3. assert that the `path` is a `str` and exists
4. (Optional) assert that the `path` is the string `api.github… etc.`
5. assert that the call to `make_request` is made with `patch`
6. assert that the return is an HTTP response.

and that's all. there's nothing else to test here. 

ok now let's go to `process_issue_action` in the case of 

**Test Condition Input**: `if issue.action == 'opened' and scope == 'public':`

1. Test 1 (Success)
    - when `issue.tag_as_public()` returns a 200 Response (mocking)
        -  `assert process_issue_action(issue).data == 'gracias, amigo.'`
        - `assert process_issue_action(issue).status_code == '200'`

2. Test 2 (Failure)
    - when `issue.tag_as_public()` returns an HTTPError (mocking)
        - `assert 'public:opened labels failed' in caplog.text`
        - `assert process_issue_action(issue).data == 'oops'`
        - `assert process_issue_action(issue).status_code == 400`

And **that's all** to be tested :) really

Now the `r.raise_for_status()` is **not** part of `process_issue_action`, so it has to be tested only at the level of `make_request` tests.

With the same logic. What are the input possible, what are the outcomes possible for this functions. Probably mocking `proxy_request`  just for the tests about `make_request`

Doing like this will make everything clearer more isolated and testing one thing at a time with less complicated tests. Everything will be more readable. :) 

* Conditions, Inputs, Outputs for each functions :) 


## about testing logging: [caplog](https://docs.pytest.org/en/latest/logging.html#caplog-fixture)

(Don't feel bad, I just discovered it now when I was trying to find out how to do it. )

```python
import logging
LOGGER = logging.getLogger(__name__)

def test_blahblah(caplog):
     caplog.set_level(logging.INFO):
     # Here all the code to set the right condition such as scope and issue value
     process_issue_action(issue)
     assert 'The log output…' in caplog.text
```

Reply:
@miketaylr Here a simple example, I made today:

```python
import logging
LOGGER = logging.getLogger(__name__)


def see(animal):
    if animal == 'snake':
        logging.warning('danger!')
    else:
        logging.info('everything is fine')
    return f'Animal : {animal}'


logging.basicConfig(level=logging.DEBUG)
see('snake')
see('grasshoper')


def test_voir(caplog):
    caplog.set_level(logging.INFO)
    voir('snake')
    assert 'danger' in caplog.text
    voir('grasshoper')
    assert 'everything is fine' in caplog.text
```

Then

```
% python test_deleteme.py
WARNING:root:dangereux
INFO:root:tout va bien
```

And 

```
% COLUMNS=40 pytest test_deleteme.py
========= test session starts ==========
platform darwin -- Python 3.8.5, pytest-5.4.3, py-1.8.1, pluggy-0.13.1
rootdir: /Users/karl/code/webcompat.com-karlcow, inifile: pytest.ini
plugins: mock-3.2.0
collected 1 item                       

test_deleteme.py .               [100%]

========== 1 passed in 0.03s ===========
```

Reply:
ah! so we can actually test that we're logging things too. very nice.

Reply:
Thanks for the "think out loud" session here. 

One slight correction - `tag_as_public` doesn't return a HTTP response, it doesn't return anything. Maybe we want it to later? But we're relying on HTTPErrors being thrown as a way to react, rather than check the rv. `make_request` does return an HTTP response, but it isn't returned outside of `tag_as_public`. **But**, that doesn't change what you are saying in terms of how to test `process_action`.

Reply:
> One slight correction - `tag_as_public` doesn't return a HTTP response, it doesn't return anything. 

correct. Thanks.